HopsanGUI/LogVariable.cpp

```diff
@@ -1774,6 +1774,114 @@ QVector<double> VectorVariable::absOfData() const
+    {
+        sindata[i] = qSin((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return sindata;
+}
+
+QVector<double> VectorVariable::cosOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> cosdata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        cosdata[i] = qCos((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return cosdata;
+}
+
+QVector<double> VectorVariable::tanOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> tandata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        tandata[i] = qTan((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return tandata;
+}
+
+QVector<double> VectorVariable::asinOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> asindata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        asindata[i] = qAsin((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return asindata;
+}
+
+QVector<double> VectorVariable::acosOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> acosdata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        acosdata[i] = qAcos((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return acosdata;
+}
+
+QVector<double> VectorVariable::atanOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> atandata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        atandata[i] = qAtan((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return atandata;
+}
+
+QVector<double> VectorVariable::expOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> expdata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        expdata[i] = qExp((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return expdata;
+}
+
+QVector<double> VectorVariable::logOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> logdata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        logdata[i] = qLn((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return logdata;
+}
+
+QVector<double> VectorVariable::sqrtOfData() const
+{
+    QVector<double> *pData = mpCachedDataVector->beginFullVectorOperation();
+    QVector<double> sqrtdata(pData->size());
+    for (int i=0; i<pData->size(); ++i)
+    {
+        sqrtdata[i] = qSqrt((*pData)[i]);
+    }
+    mpCachedDataVector->endFullVectorOperation(pData);
+    return sqrtdata;
+}
+
```

Comment:
Since all of these blocks are the same, a help function can be used, and the unique part (the math function) could be sent in as a lambda or as a template parameter in the help function. If you have the time, otherwise OK. But I guess that the existing functions are also like this.

Reply:
Good point, I replaced all the single-argument function implementations with a map of function pointers.

Reply:
I was referring to the ones in LogVariable, but sure the HCOM functions were also nearly identical

You could do something like:


`template <typename Function>`
` QVector<double> cleverFunctionName(CachedDataVectorType& pCachedDataVector, Function func )`
`    QVector<double> *pData = pCachedDataVector->beginFullVectorOperation();`
`    QVector<double> returndata(pData->size());`
`    for (int i=0; i<pData->size(); ++i)`
`    {`
`          returndata[i] = func((*pData)[i]);`
`    }`
`    pCachedDataVector->endFullVectorOperation(pData);`
`    return returndata;`

Then call with:
`return cleverFunctionName(pCachedDataVector, qCos)`

Hopefully I got this right, just writing out of memory

Reply:
Actually my commit deals with both LogVariable functions and scalar math functions, since each HCOM function call requires both. They first try to find a vector with specified name, and if it fails they look for a scalar. Hence, each function call needs two function pointer, one for the vector operation and one scalar. 

Also, I decided to use cmath functions intead of Qt equivalents, since the Qt functions have different return types while cmath functions all return double (for example, Qt's "floor" function returns an integer).

Reply:
I also implemented your suggestion, so that all "...OfData()" methods in LogVariable class are replaced with a single template function.

Reply:
Yes cmath is better in this case :)